samplers/aws/xray/client.go

```diff
@@ -0,0 +1,73 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+)
+
+type xrayClient struct {
+	// http client for sending unsigned proxied requests to the collector
+	httpClient *http.Client
+
+	endpoint *url.URL
+}
+
+// newClient returns an HTTP client with proxy endpoint
+func newClient(d string) *xrayClient {
+	endpoint := "http://" + d
+
+	endpointURL, err := url.Parse(endpoint)
+	if err != nil {
+		globalLogger.Error(err, "unable to parse endpoint from string")
+	}
+
+	return &xrayClient{
+		httpClient: &http.Client{},
+		endpoint:   endpointURL,
+	}
+}
+
+// getSamplingRules calls the collector(aws proxy enabled) for sampling rules
+func (p *xrayClient) getSamplingRules(ctx context.Context) (*getSamplingRulesOutput, error) {
+	statisticsByte, err := json.Marshal(getSamplingRulesInput{})
+	if err != nil {
+		return nil, err
+	}
+	body := bytes.NewReader(statisticsByte)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.endpoint.String()+"/GetSamplingRules", body)
+	if err != nil {
+		return nil, fmt.Errorf("xray client: failed to create http request: %w", err)
+	}
+
+	output, err := p.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("xray client: unable to retrieve sampling settings: %w", err)
+	}
+	defer output.Body.Close()
+
+	var samplingRulesOutput *getSamplingRulesOutput
+	if err := json.NewDecoder(output.Body).Decode(&samplingRulesOutput); err != nil {
```

Comment:
This seems to be passing a pointer to a pointer - I'm surprised this seems to actually work , but think you don't need `&`, alternatively you might not need to return a pointer here since it should get moved to the caller

Reply:
I tried removing `&` and it throwed the unmarshal error.

Reply:
I believe json.Unmarshal() handles this correctly, but it would be nicer to see 
```
var samplingRulesOutput getSamplingRulesOutput
```
i.e., drop the `*`, not the `&`